Initial commit of IronSFTP changes.

BobWall23 opened the pull request:
Still unit test and regression test work, but code changes ready to review.

BobWall23 committed:
Initial commit of IronSFTP changes.

BobWall23 committed:
Remove ironsftp-server, ironscp.

No longer need changes to SFTP server for first release. Reverted to
production server. Removed changes from client that required server
changes to access. Removed ironsftp-server from Makefile.in. Also
removed ironscp, since it isn't updated yet.

zmre commented on Makefile.in:
```diff
@@ -62,6 +64,17 @@ LDFLAGS=-L. -Lopenbsd-compat/ @LDFLAGS@
 MANFMT=@MANFMT@
 
+# *** ICL Modification ***
+SODIUM_VER = 1.0.10
```
Just a quick note that they are on 1.0.11 now. Homebrew has that if you upgrade.

zmre commented on Makefile.in:
```diff
@@ -142,12 +155,15 @@ FIXPATHSCMD	= $(SED) $(PATHSUBS)
 
-all: $(CONFIGFILES) $(MANPAGES) $(TARGETS)
+#all: $(CONFIGFILES) $(MANPAGES) $(TARGETS)   FIXME
+all: $(IRON_TARGETS)
```
Do you need an issue to track outstanding todos?  There's a fixme here. Looks like man pages and config files aren't being built. Not sure what else. Not sure if it matters. But the commented line should be deleted.

zmre commented on Makefile.in:
```diff
@@ -160,6 +176,19 @@ libssh.a: $(LIBSSH_OBJS)
+ironscp$(EXEEXT):  $(LIBCOMPAT) scp.o libssh.a
+	$(LD) -o $@ scp.o $(LDFLAGS) -lssh -lopenbsd-compat $(LIBS)
+
+irongpg$(EXEEXT): iron-gpg-test.o sftp-common.o libssh.a $(LIBCOMPAT)
```
What is irongpg?  Is that something I can use to inspect files?  I just tried `make irongpg` but the build failed.

zmre commented on iron-gpg.h:
```diff
@@ -0,0 +1,44 @@
+
+#define COMMENT_MAX				128		//  Max # bytes in the comment on a public SSH key
+
+#define GPG_MAX_KEY_SIZE		512		//  # bytes a secret or public key can occupy
```
This seems really big for Curve25519, which would be 32. Is this including other things? Or allowing for RSA? If RSA, it is pretty small.

BobWall23 replied:
For both. It covers up to RSA-4096. I guess I could double it in case of RSA-8192?

zmre replied:
4096 is probably big enough. But are we using RSA? Allowing it?

BobWall23 committed:
Modify unittests for iron-gpg.

Make the iron-gpg.c unit tests fit into the framework.
Fix comments on compute_gpg_s2k_key.

BobWall23 committed:
Removed asserts, replaced by return codes

Removed the asserts from the code, replaced by returning negative values
from functions. Check return values where functions are called.
Also added new utility regress/enc-dec-file that will use the code to
encode a file into a .iron file, decrypt it, and compare the output to
the original file. Will end up getting used in regression test scripts.

zmre commented on sftp-client.c:
```diff
@@ -1598,6 +1635,14 @@ do_upload(struct sftp_conn *conn, const char *local_path,
+		return(-1);
+	}
+	/* *** */
+
```
Is local_path edited to add the .iron suffix inside write_gpg_encrypted_file?  I must have seen that.

BobWall23 replied:
Yeah - happens in open_encrypted_output_file.

zmre commented on sftp.c:
```diff
@@ -210,10 +219,10 @@ static const struct CMD cmds[] = {
 	{ "symlink",	I_SYMLINK,	REMOTE	},
 	{ "version",	I_VERSION,	NOARGS	},
 	{ "!",		I_SHELL,	NOARGS	},
-	{ "?",		I_HELP,		NOARGS	},
```
Why remove the question mark for help?

BobWall23 replied:
Editing while asleep? Restored.

zmre commented on sftp.c:
```diff
@@ -934,7 +971,14 @@ do_globbed_ls(struct sftp_conn *conn, const char *path,
+				mprintf("  %-*s", (colspace - 2), fname);
+			}
+			/* *** */
 			if (c >= columns) {
```
I'm looking at this without a lot of context, but I think this is the third (fourth?) place where the lock icon is prefixed. Presumably this happens in download progress and in directory listings. When else do we do that prefixing?

BobWall23 replied:
It is in woefully many places, because there are woefully many places that print out the same thing. One is in the progress meter, one is in sftp-common.c:ls_file, and three are in sftp.c. Two are in do_ls_dir, where in two different code paths it generates the listing, but one is aligned into columns and the other is a long listing format, and the third is in do_globbed_ls. It looks like I could at least put that into a common routine - got sloppy.

BobWall23 replied:
Decided not to. One prints to a string, one needs to peel out the dirname of the file then add the icon, etc. They are different enough it's not worth it.

zmre commented:
Alright, I made it through the whole review.  Turning my attention to grabbing the latest and doing some QA.

BobWall23 committed:
Updates following code review, fix bugs.

Replaced several magic numbers in code with constants.
Fixed bug that caused write past end of buffer (htonl problem).

BobWall23 committed:
Reorder ICL init

Moved the ICL init code to after the SSH connection was established, to
avoid weird conflicts over using stdin between the SSH connection and
the ICL key gen code.

BobWall23 committed:
Better msg, error handling for keygen

Added error handling and text output to help users understand what is
happening when they start ironsftp and don't have GPG keys yet, and to
report errors if they occur.

BobWall23 committed:
Added .gitignore

BobWall23 committed:
Removed mktemp, memmem.

Worked on OS/X, did wildly unpredictable stuff on Linux. Replaced with alternative

BobWall23 committed:
Rewrote passphrase generation to use ssh-agent

Since it is not possible to just retrieve the SSH RSA secret key from
ssh-agent, rewrote passphrase generation to hash the public parameters
(n and e) of the RSA key, then ask the ssh-agent to sign them (which
requires the secret key). If the agent is not running, or it can't sign
(probably because the key hasn't been added to the agent), retrieves the
secret key from the id_rsa file and signs the hash.

This required a fairly big refactor of code, since the public RSA key
has two parameters (n and e), and previously, I was only passing around
the secret parameter d.

BobWall23 committed:
Fixed problems with passphrase generation.

Refactored to fetch public keys and private keys separately, so that
when downloading a file, it wouldn't try to access the secret key file
before it figured out that the incoming file wasn't encrypted.

In the process, discovered problem with passphrase generation. The
passphrase generated was different if ssh-agent was running than if it
was not, and we read the secret key file directly then did the RSA
signature. Turns out that the system ssh-agent ignored the requested
"rsa-sha2-256" algorithm and just did "ssh-rsa", which is a SHA1 hash.
Finally gave up and just requested "ssh-rsa" from both ssh-agent and
local signing, so they should always match.

BobWall23 committed:
Update configure.ac to check for sodium hdr, lib

BobWall23 committed:
Refactor build, ICL modifications.

Reworked changes to existing code to be inside #ifdef IRONCORE, so that
it is possible to build the original sftp and sftp-server executables
along with ironsftp (in part to facilitate testing). Corresponding
changes to the Makefile to define IRONCORE when compiling files to build
iron executables.

Also added warning if a .iron file is downloaded but can't be encrypted
(because the file is not encrypted, or it isn't encrypted to the current
user.

BobWall23 committed:
Don't display lock icon for non-.iron files.

BobWall23 committed:
Fetch <file>.iron if <file> doesn't exist.

If the user tries to get a file and it doesn't exist on the server, if
the file doesn't end in ".iron", append ".iron" and try to get that file.

BobWall23 committed:
Append .iron to filename when uploading

Got most cases, but if you did "put filea fileb", it didn't append
".iron" to "fileb". Fixed.

BobWall23 committed:
Change naming of decrypted files.

Instead of restoring the output of decryption to a file with the name of
the original that was encrypted, just strip the .iron extension from (if
there is one), or add a .decrypt extension to the name of the downloaded
file (if it was encrypted). It does still output an info message if the
output file name is different from the name of the file that was
originally encrypted.

BobWall23 committed:
Code review fixes, copying RSA key files

Copy the id_rsa and id_rsa.pub files to id_rsa.iron and id_rsa.iron.pub
when we generate keys.

BobWall23 committed:
Init vars to keep gcc happy.

BobWall23 committed:
Unbreak errors from last commit

BobWall23 committed:
Display lock icon on files being uploaded.

If you do "put abc", ironsftp was displaying abc in the progress meter,
and because the file name did not end in .iron, it didn't display the
lock icon. Fixed so that it displays "abc.iron" as the name of the file
being uploaded, which displays the lock icon.

Also restat'd the file after it was encrypted, so the file size properly
matched the file that was actually being uploaded.

BobWall23 committed:
Tweak handling of .pubkey file.

Make the process of adding lines to the .pubkey file when new keys are
generated more robust, without making assumptions on the maximum length
of the line.

BobWall23 committed:
Modified handling of file names.

Now, write_gpg_encrypted_file generates an output file name by
appending .iron to the input file. If that file already exists, it
generates a file with a .iron extension using mkstemps.
write_gpg_decrypted_file strips the .iron from the input file and opens
that name as the output file. It will overwrite the file if it exists.
An ironsftp put will always append .iron to the remote file name (if it
doesn't already have it). If the remote file exists, prompts to confirm
overwrite.
An ironsftp put will always append .iron to the local file name to form
the download file name. If that name exists, prompts to confirm
overwrite. It also prompts to confirm overwrite of the file without the
.iron extension. After the .iron file is downloaded, if it was not
encrypted, renames it from the .iron file to the output file name.

BobWall23 committed:
Fixed compile error

leifwickland commented on progressmeter.c:
```diff
@@ -160,11 +163,25 @@ refresh_progress_meter(void)
+#ifdef IRONCORE
+#define LOCK_PREFIX_EXTRA_CHARS 4	//  For the return character and the first 3 of 4 bytes of the lock icon,
+	   								//	none of which take any space on the screen.
+		if (iron_extension_offset(file) >= 0) {
```
If the file name consists of nothing more than the `.iron` extension do you want to go down this path? I really have no idea what the right thing to do in that case would be, but it certainly seems degenerate.

BobWall23 replied:
True. In a lot of the other places where I check for the extension, I ignore it if the offset is 0. I'll make this consistent.

leifwickland commented on regress/enc-dec-file.c:
```diff
@@ -0,0 +1,61 @@
+	
+	//  Create a <infile>.iron file, so that write_gpg_encrypted_file will generate a different output file.
+	char tmp_fname[PATH_MAX + 1];
+	sprintf(tmp_fname, "%s%s", argv[1], IRON_SECURE_FILE_SUFFIX);
```
Should this be a `snprintf` to avoid exceeding `tmp_fname`? Should there be a check to detect that the result didn't fit in `PATH_MAX`?

BobWall23 replied:
Got sloppy - it was only a test file. But added a check for length of argv[1] at start of execution.

leifwickland commented on sftp-common.c:
```diff
@@ -257,3 +271,35 @@ ls_file(const char *name, const struct stat *st, int remote, int si_units)
+ *  @return int Offset of suffix in fname, or -1 if fname doesn't end with suffix
+ */
+int
+iron_extension_offset(const char * name)
```
It seems like it'd be simpler and not particularly worse performing to find the end of the string and then check if  `n-IRON_SECURE_FILE_SUFFIX_LEN >= 0` and `name + (n-IRON_SECURE_FILE_SUFFIX_LEN)` was `IRON_SECURE_FILE_SUFFIX`.

:shrug:

I might be out of practice juggling  chainsaws.

BobWall23 replied:
Well played, sir.

leifwickland commented on sftp-client.c:
```diff
@@ -694,6 +699,16 @@ do_mkdir(struct sftp_conn *conn, const char *path, Attrib *a, int print_flag)
+	 *  that store sharing info.
+	 */
+	if (iron_extension_offset(path) >= 0) {
+		error("Directory name suffix \"%s\" reserved for use on the server", IRON_SECURE_FILE_SUFFIX);
```
Do you think it'd be a bad idea to put the original `path` into the error message? Something like:
`error("Directory name suffix \"%s\" reserved for use on the server. \"%s\ is not a permitted directory name.", IRON_SECURE_FILE_SUFFIX, name);

leifwickland commented on sftp-client.c:
```diff
@@ -1140,6 +1155,20 @@ send_read_request(struct sftp_conn *conn, u_int id, u_int64_t offset,
+		fgets(line, sizeof(line), stdin);
+	}
+	return yesno;
+}
```
It seems like you ought to include the logic to check if `yesno` is `Y` or `y` in here and just return a boolean-ish. It sure looks like that's how you actually use it. Obviously a rename would be in order at that point.

BobWall23 replied:
Agreed - more better. Now `static int get_user_confirmation()`

leifwickland commented on sftp-client.c:
```diff
@@ -1193,6 +1222,41 @@ do_download(struct sftp_conn *conn, const char *remote_path,
 
+#ifdef IRONCORE
+	if (strlen(local_path) > PATH_MAX - IRON_SECURE_FILE_SUFFIX_LEN) {
+		error("Local path name %s too long.", local_path);
```
Typically above this point in the PR you had been putting path names inside of `"..."` within error messages. In this file it appears you're not. Are you happy with that inconsistency? 
Oh, a bit further down I see that you're quoting again. There's some level of inconsistency which you may want to address.

BobWall23 replied:
Agreed - consistentified.

leifwickland commented on sftp-client.c:
```diff
@@ -1193,6 +1222,41 @@ do_download(struct sftp_conn *conn, const char *remote_path,
+		return(-1);
+	}
+	const char * output_path;
+	char iron_name[PATH_MAX + IRON_SECURE_FILE_SUFFIX_LEN + 1];
```
Is it a good idea to create paths which are longer than `PATH_MAX`? I suspect other apps and/or the OS will be unhappy about that.

leifwickland commented on sftp-client.c:
```diff
@@ -1404,6 +1468,29 @@ do_download(struct sftp_conn *conn, const char *remote_path,
+		 *  it should have a ".iron" suffix. If the file can be decrypted, the output file will have the ".iron"
+		 *  stripped. If the file wasn't encrypted, we should try to rename without the .iron suffix.
+		 */
+		char dec_fname[PATH_MAX + 1];
```
It seems like `PATH_MAX` is supposed to include the null.

See the comment on https://github.com/torvalds/linux/blob/master/include/uapi/linux/limits.h#L12

If you agree with that, it seems like a bunch of other places should be updated too.

BobWall23 replied:
Yup.

leifwickland commented on sftp-client.c:
```diff
@@ -1193,6 +1222,41 @@ do_download(struct sftp_conn *conn, const char *remote_path,
 	attrib_clear(&junk); /* Send empty attributes */
 
+#ifdef IRONCORE
+	if (strlen(local_path) > PATH_MAX - IRON_SECURE_FILE_SUFFIX_LEN) {
```
Should you be checking if the file name plus the iron extension exceeds `NAME_MAX`?

leifwickland commented on sftp-common.c:
```diff
@@ -245,6 +249,16 @@ ls_file(const char *name, const struct stat *st, int remote, int si_units)
+	/*  If file has .iron extension, prefix with lock icon  */
+	int offset = iron_extension_offset(name);
+	char nbuf[512];
+	char * icon_prefix = (offset > 0) ? IRON_LOCK_ICON : "  ";
```
Most other places you allow the file name to consist of only the iron extension. Here you don't. Merely pointing out an inconsistency.

leifwickland commented on sftp-common.c:
```diff
@@ -245,6 +249,16 @@ ls_file(const char *name, const struct stat *st, int remote, int si_units)
+#ifdef IRONCORE
+	/*  If file has .iron extension, prefix with lock icon  */
+	int offset = iron_extension_offset(name);
+	char nbuf[512];
```
Why is this large enough to hold the name plus the prefix?

leifwickland commented on sftp-common.c:
```diff
@@ -245,6 +249,16 @@ ls_file(const char *name, const struct stat *st, int remote, int si_units)
+	char nbuf[512];
+	char * icon_prefix = (offset > 0) ? IRON_LOCK_ICON : "  ";
+	sprintf(nbuf, "%s%s", icon_prefix, name);
+	name = nbuf;
```
I don't see what effect this has. `name` isn't used below this point, is it? 
_UPDATE:_ Sorry. I didn't see that section I needed to expand where it is used. My bad.

leifwickland commented on sftp.c:
```diff
@@ -672,6 +717,21 @@ process_get(struct sftp_conn *conn, const char *src, const char *dst,
 
+#ifdef IRONCORE
+char *
+iron_append(const char *p1)
```
Should you to anything to ensure that `p1` doesn't end with a `/`? Suffixing that would lead to undesirable results.

BobWall23 replied:
Turns out I no longer use the function.

leifwickland commented on sftp.c:
```diff
@@ -604,11 +624,36 @@ process_get(struct sftp_conn *conn, const char *src, const char *dst,
+			err = -1;
+		} else if (iron_extension_offset(src) < 0) {
+			//  If the requested file didn't have a .iron extension, add one and try again
+			char * iron_name = xmalloc(strlen(abs_src) + IRON_SECURE_FILE_SUFFIX_LEN + 1);
```
Can you reuse `iron_append` here to create `iron_name`?

leifwickland commented on sftp.c:
```diff
@@ -736,6 +796,14 @@ process_put(struct sftp_conn *conn, const char *src, const char *dst,
 		free(tmp);
+#ifdef IRONCORE
+		/*  Append .iron extension to destination file name  */
+		char iron_name[PATH_MAX + IRON_SECURE_FILE_SUFFIX_LEN + 1];
```
Can you reuse iron_append here to create iron_name?

leifwickland replied:
Ditto previous concerns about exceeding `PATH_MAX`.

leifwickland commented on sftp.c:
```diff
@@ -810,6 +878,10 @@ do_ls_dir(struct sftp_conn *conn, const char *path,
 		free(tmp);
 
+#ifdef IRONCORE
+		m += 2;  /*  Add space for lock icon or two spaces if not required  */
```
Suggestion: Consider defining constants for this and the `"  "` not-locked string in the same area as the locked string.

BobWall23 replied:
👍

leifwickland commented on sftp.c:
```diff
@@ -847,10 +919,31 @@ do_ls_dir(struct sftp_conn *conn, const char *path,
+
+				int len = fname_start - d[n]->longname;
+				mprintf("%-*.*s", len, len, d[n]->longname);
+				char * icon_prefix = (offset > 0) ? IRON_LOCK_ICON : "  ";
```
Ditto comment about inconsistency around `offset > 0` vs `>= 0`

BobWall23 replied:
Should all be consistently `>0` now, except for one case which is commented to explain difference.

leifwickland commented on sftp.c:
```diff
@@ -1387,9 +1494,21 @@ parse_args(const char **cpp, int *ignore_errors, int *aflag,
+#ifdef IRONCORE
+	case I_ADD_RCPT:
+	case I_RM_RCPT:
 		if ((optidx = parse_no_flags(cmd, argv, argc)) == -1)
```
It seems like we could safely assume any return `<= 0` is bad here since you're about to index into `argv` with it

leifwickland commented on sftp.c:
```diff
@@ -1387,9 +1494,21 @@ parse_args(const char **cpp, int *ignore_errors, int *aflag,
 		if ((optidx = parse_no_flags(cmd, argv, argc)) == -1)
 			return -1;
+		/* Login is required */
+		if (argc - optidx < 1) {
```
Nitpick: I'd normally write that as `optidx >= argc`. Feel free to leave it. I just had to think hard about what you were doing.

BobWall23 replied:
Cut and paste of some previous code. But that doesn't excuse it.

leifwickland commented:
My condolences. That was painful to read and it was some damned fine C. That had to be wickedly painful to write.

BobWall23 committed:
Code review changes, max size of recipient list.

Various changes from code reviews.
Changed recipient list so it maxes out at 10 users (plus the current
user), instead of growing without bound. Support for more sharing will
be part of paid offering.
